{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Structural analysis of discontinuous concepts\n",
    "\n",
    "The purpose of this analysis notebook is to make an in-depth analysis of structures. This makes the basis for a large portion of the corresponding section in the final thesis paper."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Coordinated concepts\n",
    "\n",
    "Coordinations are the only thing marked in GENIA while CRAFT also contains other types. Therefore, GENIA will be used primarily in this part. The variable `corpus_name` can be set to either `genia` or `craft` to run with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import utils  # a python module in the same dir as the notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "corpus_name = 'genia'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, let's load the corpus, then extract concepts and their constituent structures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading GENIA corpus ... NOTE: 13 files cannot get Constituent annotations!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1599/1599 [00:15<00:00, 105.69it/s]\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "os.chdir(utils.ROOT)  # get to the root directory of the project\n",
    "\n",
    "from datautils import dataio, annotations as anno\n",
    "\n",
    "# load the corpora\n",
    "corpus = dataio.load_corpus(corpus_name)\n",
    "\n",
    "# not all docs in genia get Constituent annotations; if so, leave them out\n",
    "if corpus_name.lower() == 'genia':\n",
    "    corpus = [doc for doc in corpus if doc.get_annotations(anno.Constituent)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "concept_consts = [(concept, doc.get_annotations_at(concept.span, anno.Constituent)[0])\n",
    "                  for doc in corpus\n",
    "                  for concept in doc.get_annotations(anno.DiscontinuousConcept)\n",
    "                  if len(concept.get_spanned_tokens()) < 10]\n",
    "\n",
    "# for craft, keep only the ones with a conjunction in them\n",
    "temp_list = []\n",
    "for concept, const in concept_consts:\n",
    "    present_pos = {t.mapped_pos() for t in const.get_tokens()}\n",
    "    if 'c' in present_pos:\n",
    "        temp_list.append((concept, const))\n",
    "concept_consts = temp_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each constituent can spit out a treebank structure. Let's do that and put it all in dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def skipped_structure(const, allowed_tokens):\n",
    "    \"\"\"Returns  a treebank structure of a Constituent with a # symbol in front of\n",
    "    all tokens that are in allowed_tokens\"\"\"\n",
    "    # note: craft constituents often have labels like NP-SBJ; keep only the first part\n",
    "    return '(' + const.label.split('-')[0] + ' ' + ' '.join(\n",
    "        skipped_structure(c, allowed_tokens) if isinstance(c, anno.Constituent)\n",
    "        else c.mapped_pos() if c not in allowed_tokens\n",
    "        else '#' + c.mapped_pos()\n",
    "        for c in const.constituents\n",
    "    ) + ')'\n",
    "\n",
    "def unlabeled_structure(const, allowed_tokens):\n",
    "    \"\"\"Returns a simplified treebank structure of a Constituent with no constituent labels\n",
    "    and with a # symbol in front of all tokens that are in allowed_tokens\"\"\"\n",
    "    return_string = '(' + ' '.join(\n",
    "        unlabeled_structure(c, allowed_tokens) if isinstance(c, anno.Constituent)\n",
    "        else c.mapped_pos() if c not in allowed_tokens\n",
    "        else '#' + c.mapped_pos()\n",
    "        for c in const.constituents\n",
    "        if not contains_empty_token(c)\n",
    "    ) + ')'\n",
    "    # remove all cases of single token constituents, e.g. (n)\n",
    "    return re.sub('\\(+(#?.)\\)+', r'\\1', return_string)\n",
    "\n",
    "def contains_empty_token(const):\n",
    "    if not isinstance(const, anno.Constituent):\n",
    "        return False\n",
    "    for c in const.constituents:\n",
    "        if isinstance(c, anno.Token) and c.mapped_pos() == 'Ø':\n",
    "            return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "\n",
    "full_structures = defaultdict(list)\n",
    "skip_structures = defaultdict(list)\n",
    "collapsed_full_structures = defaultdict(list)\n",
    "collapsed_skip_structures = defaultdict(list)\n",
    "\n",
    "for concept, const in concept_consts:\n",
    "    struct = const.structure()  \n",
    "    full_structures[struct].append(const)\n",
    "    \n",
    "    skip_struct = skipped_structure(const, concept.get_tokens())\n",
    "    skip_structures[skip_struct].append(const)\n",
    "    \n",
    "    collapsed_full = unlabeled_structure(const, {})\n",
    "    collapsed_full_structures[collapsed_full].append(const)\n",
    "    \n",
    "    collapsed_skip = unlabeled_structure(const, concept.get_tokens())\n",
    "    collapsed_skip_structures[collapsed_skip].append(const)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_dcs = len(concept_consts)\n",
    "\n",
    "dfs = {}\n",
    "\n",
    "for name, structure_dict in [('full', full_structures),\n",
    "                             ('skip', skip_structures),\n",
    "                             ('col_full', collapsed_full_structures),\n",
    "                             ('col_skip', collapsed_skip_structures)]:\n",
    "    \n",
    "    data_dict = {'struct': [], 'pos-seq': [], 'count': [], '%': [], 'example': []}\n",
    "    \n",
    "    for struct, sample in structure_dict.items():\n",
    "        data_dict['struct'].append(struct)\n",
    "        data_dict['pos-seq'].append(''.join(c for c in struct if c.islower() or c in ','))\n",
    "        data_dict['count'].append(len(sample))\n",
    "        data_dict['%'].append(round(len(sample) / n_dcs * 100, 2))\n",
    "        data_dict['example'].append(sample[0].get_covered_text())\n",
    "        \n",
    "    dfs[name] = pd.DataFrame(data_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>struct</th>\n",
       "      <th>pos-seq</th>\n",
       "      <th>count</th>\n",
       "      <th>%</th>\n",
       "      <th>example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>(#n c n #n)</td>\n",
       "      <td>ncnn</td>\n",
       "      <td>74</td>\n",
       "      <td>5.21</td>\n",
       "      <td>alpha and beta subunits</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>(#a c a #n)</td>\n",
       "      <td>acan</td>\n",
       "      <td>58</td>\n",
       "      <td>4.08</td>\n",
       "      <td>lymphoid and nonlymphoid cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>(t #n c n #n)</td>\n",
       "      <td>tncnn</td>\n",
       "      <td>37</td>\n",
       "      <td>2.61</td>\n",
       "      <td>the TSG101 and FHIT genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>((#n c n) #n)</td>\n",
       "      <td>ncnn</td>\n",
       "      <td>29</td>\n",
       "      <td>2.04</td>\n",
       "      <td>H9 and CEM cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>((#a c a) #n)</td>\n",
       "      <td>acan</td>\n",
       "      <td>20</td>\n",
       "      <td>1.41</td>\n",
       "      <td>cell-cell and cell-matrix interactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>(#n c n (#n #n))</td>\n",
       "      <td>ncnnn</td>\n",
       "      <td>17</td>\n",
       "      <td>1.20</td>\n",
       "      <td>T and B cell co-cultures</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>((#n c (f #a)) #n)</td>\n",
       "      <td>ncfan</td>\n",
       "      <td>12</td>\n",
       "      <td>0.85</td>\n",
       "      <td>glucocorticoid- and Fas-mediated cytotoxicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>(#a c a (#n #n))</td>\n",
       "      <td>acann</td>\n",
       "      <td>11</td>\n",
       "      <td>0.77</td>\n",
       "      <td>Tax-expressing and HTLV-I-transformed T cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>((#a c a) #n #n)</td>\n",
       "      <td>acann</td>\n",
       "      <td>11</td>\n",
       "      <td>0.77</td>\n",
       "      <td>normal and dex-resistant CEM cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(t (#n c n) #n)</td>\n",
       "      <td>tncnn</td>\n",
       "      <td>11</td>\n",
       "      <td>0.77</td>\n",
       "      <td>the GM-CSF and IL-2 genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>(#n c (f #a) #n)</td>\n",
       "      <td>ncfan</td>\n",
       "      <td>10</td>\n",
       "      <td>0.70</td>\n",
       "      <td>TNF-alpha- or PMA-induced stimulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>(c #a c a #n)</td>\n",
       "      <td>cacan</td>\n",
       "      <td>9</td>\n",
       "      <td>0.63</td>\n",
       "      <td>both transcriptional and post-transcriptional ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>(t (#a c a) #n)</td>\n",
       "      <td>tacan</td>\n",
       "      <td>9</td>\n",
       "      <td>0.63</td>\n",
       "      <td>the sensitive and resistant patients</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>((#n c n) #n #n)</td>\n",
       "      <td>ncnnn</td>\n",
       "      <td>8</td>\n",
       "      <td>0.56</td>\n",
       "      <td>CD2 and CD28 adhesion molecules</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>(#a c n #n)</td>\n",
       "      <td>acnn</td>\n",
       "      <td>8</td>\n",
       "      <td>0.56</td>\n",
       "      <td>human and mouse cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>((#a c a) #a #n)</td>\n",
       "      <td>acaan</td>\n",
       "      <td>7</td>\n",
       "      <td>0.49</td>\n",
       "      <td>erythroid and megakaryocytic terminal maturation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>(#n c n (#n #n #n))</td>\n",
       "      <td>ncnnnn</td>\n",
       "      <td>7</td>\n",
       "      <td>0.49</td>\n",
       "      <td>Nb2 and C196 T cell lines</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>(#n d c #d)</td>\n",
       "      <td>ndcd</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>serines 32 and 36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>(#n (d c #d))</td>\n",
       "      <td>ndcd</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>Interleukin 2 and 15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>(t #a c a #n)</td>\n",
       "      <td>tacan</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>the erythroid and megakaryocytic lineages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>((c #n c n) #n)</td>\n",
       "      <td>cncnn</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>both HTLV-I and HTLV-II Tax</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>((c #a c a) #n)</td>\n",
       "      <td>cacan</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>both erythroid and megakaryocytic lineages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>(#n c (f #a) (#n #n))</td>\n",
       "      <td>ncfann</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>androgen- and antiandrogen-bound receptor comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>(c #n c n #n)</td>\n",
       "      <td>cncnn</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>both ERK1/2 and p38 cascades</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>(a #a c a #n)</td>\n",
       "      <td>aacan</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>novel immunosuppressive and antiviral therapies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>(#n n c #n)</td>\n",
       "      <td>nncn</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>peptide binding and recognition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>(a #n c n #n)</td>\n",
       "      <td>ancnn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>constitutive SLA-DR and SLA-DQ expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>(#a c a (#n #n #n))</td>\n",
       "      <td>acannn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>murine and human p40 promoter regions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>(#n (n c #n))</td>\n",
       "      <td>nncn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>genes UL49 and UL50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>((#n c (f #a)) #n #n)</td>\n",
       "      <td>ncfann</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>Sp1- and Ets-related transcription factors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>((#f n) c #n)</td>\n",
       "      <td>fncn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>interferon-alpha and -gamma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>(c #a c a (#n #n))</td>\n",
       "      <td>cacann</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>both B-cell-restricted and ubiquitous transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>(#n , n , c n #n)</td>\n",
       "      <td>n,n,cnn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>M-TAT/EPO, M-TAT/GM-CSF, and M-TAT/SCF cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>(#n c n (#a #n))</td>\n",
       "      <td>ncnan</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>kappa and lambda light-chain genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>(n , #n , c n #n)</td>\n",
       "      <td>n,n,cnn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>M-TAT/EPO, M-TAT/GM-CSF, and M-TAT/SCF cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(#n c (f #n))</td>\n",
       "      <td>ncfn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>c- and v-rel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>(#a n c #n)</td>\n",
       "      <td>ancn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>erythroid differentiation and maturation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>(#n c n #n #n)</td>\n",
       "      <td>ncnnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>c-fos and c-jun mRNA expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>(#a #n c n #n)</td>\n",
       "      <td>ancnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>murine Th1 and Th2 cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>(#a (a c #a) #n)</td>\n",
       "      <td>aacan</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>Recombinant murine and human TCF-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>((#f n) , #d , c d)</td>\n",
       "      <td>fn,d,cd</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>TRAF1, -2, or -3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>(#n c a (#n #n))</td>\n",
       "      <td>ncann</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>macrophage and epithelial cell lines</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>((#f n) , d , c #d)</td>\n",
       "      <td>fn,d,cd</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>TRAF1, -2, or -3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>(#n #n (n c #n))</td>\n",
       "      <td>nnncn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>PKC isoenzymes alpha and beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>((#n c (f #n)) #n)</td>\n",
       "      <td>ncfnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>B- and T-cell development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>((#a #n) c (n #n))</td>\n",
       "      <td>ancnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>abnormal cAMP and cAMP-phosphodiesterase activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>(t (a , #a , c a) #n)</td>\n",
       "      <td>ta,a,can</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>the erythroid, megakaryocytic, and mastocytic/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>((#f n) c #d)</td>\n",
       "      <td>fncd</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>cyclin-dependent-kinase-4 and 6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>(t (#a , a , c a) #n)</td>\n",
       "      <td>ta,a,can</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>the erythroid, megakaryocytic, and mastocytic/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>(n , #n c n #n)</td>\n",
       "      <td>n,ncnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>NF-AT, AP-1 and CLE0 sequences</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    struct   pos-seq  count     %  \\\n",
       "93             (#n c n #n)      ncnn     74  5.21   \n",
       "50             (#a c a #n)      acan     58  4.08   \n",
       "17           (t #n c n #n)     tncnn     37  2.61   \n",
       "83           ((#n c n) #n)      ncnn     29  2.04   \n",
       "62           ((#a c a) #n)      acan     20  1.41   \n",
       "138       (#n c n (#n #n))     ncnnn     17  1.20   \n",
       "187     ((#n c (f #a)) #n)     ncfan     12  0.85   \n",
       "66        (#a c a (#n #n))     acann     11  0.77   \n",
       "137       ((#a c a) #n #n)     acann     11  0.77   \n",
       "4          (t (#n c n) #n)     tncnn     11  0.77   \n",
       "42        (#n c (f #a) #n)     ncfan     10  0.70   \n",
       "89           (c #a c a #n)     cacan      9  0.63   \n",
       "197        (t (#a c a) #n)     tacan      9  0.63   \n",
       "73        ((#n c n) #n #n)     ncnnn      8  0.56   \n",
       "169            (#a c n #n)      acnn      8  0.56   \n",
       "214       ((#a c a) #a #n)     acaan      7  0.49   \n",
       "9      (#n c n (#n #n #n))    ncnnnn      7  0.49   \n",
       "76             (#n d c #d)      ndcd      6  0.42   \n",
       "74           (#n (d c #d))      ndcd      6  0.42   \n",
       "207          (t #a c a #n)     tacan      6  0.42   \n",
       "185        ((c #n c n) #n)     cncnn      6  0.42   \n",
       "236        ((c #a c a) #n)     cacan      6  0.42   \n",
       "104  (#n c (f #a) (#n #n))    ncfann      5  0.35   \n",
       "55           (c #n c n #n)     cncnn      5  0.35   \n",
       "517          (a #a c a #n)     aacan      5  0.35   \n",
       "49             (#n n c #n)      nncn      5  0.35   \n",
       "245          (a #n c n #n)     ancnn      4  0.28   \n",
       "82     (#a c a (#n #n #n))    acannn      4  0.28   \n",
       "333          (#n (n c #n))      nncn      4  0.28   \n",
       "91   ((#n c (f #a)) #n #n)    ncfann      4  0.28   \n",
       "72           ((#f n) c #n)      fncn      4  0.28   \n",
       "241     (c #a c a (#n #n))    cacann      4  0.28   \n",
       "208      (#n , n , c n #n)   n,n,cnn      4  0.28   \n",
       "110       (#n c n (#a #n))     ncnan      4  0.28   \n",
       "209      (n , #n , c n #n)   n,n,cnn      4  0.28   \n",
       "3            (#n c (f #n))      ncfn      4  0.28   \n",
       "297            (#a n c #n)      ancn      4  0.28   \n",
       "128         (#n c n #n #n)     ncnnn      3  0.21   \n",
       "393         (#a #n c n #n)     ancnn      3  0.21   \n",
       "413       (#a (a c #a) #n)     aacan      3  0.21   \n",
       "64     ((#f n) , #d , c d)   fn,d,cd      3  0.21   \n",
       "140       (#n c a (#n #n))     ncann      3  0.21   \n",
       "65     ((#f n) , d , c #d)   fn,d,cd      3  0.21   \n",
       "356       (#n #n (n c #n))     nnncn      3  0.21   \n",
       "71      ((#n c (f #n)) #n)     ncfnn      3  0.21   \n",
       "122     ((#a #n) c (n #n))     ancnn      3  0.21   \n",
       "213  (t (a , #a , c a) #n)  ta,a,can      3  0.21   \n",
       "166          ((#f n) c #d)      fncd      3  0.21   \n",
       "212  (t (#a , a , c a) #n)  ta,a,can      3  0.21   \n",
       "451        (n , #n c n #n)    n,ncnn      3  0.21   \n",
       "\n",
       "                                               example  \n",
       "93                             alpha and beta subunits  \n",
       "50                      lymphoid and nonlymphoid cells  \n",
       "17                           the TSG101 and FHIT genes  \n",
       "83                                    H9 and CEM cells  \n",
       "62              cell-cell and cell-matrix interactions  \n",
       "138                           T and B cell co-cultures  \n",
       "187      glucocorticoid- and Fas-mediated cytotoxicity  \n",
       "66       Tax-expressing and HTLV-I-transformed T cells  \n",
       "137                 normal and dex-resistant CEM cells  \n",
       "4                            the GM-CSF and IL-2 genes  \n",
       "42               TNF-alpha- or PMA-induced stimulation  \n",
       "89   both transcriptional and post-transcriptional ...  \n",
       "197               the sensitive and resistant patients  \n",
       "73                     CD2 and CD28 adhesion molecules  \n",
       "169                              human and mouse cells  \n",
       "214   erythroid and megakaryocytic terminal maturation  \n",
       "9                            Nb2 and C196 T cell lines  \n",
       "76                                   serines 32 and 36  \n",
       "74                                Interleukin 2 and 15  \n",
       "207          the erythroid and megakaryocytic lineages  \n",
       "185                        both HTLV-I and HTLV-II Tax  \n",
       "236         both erythroid and megakaryocytic lineages  \n",
       "104  androgen- and antiandrogen-bound receptor comp...  \n",
       "55                        both ERK1/2 and p38 cascades  \n",
       "517    novel immunosuppressive and antiviral therapies  \n",
       "49                     peptide binding and recognition  \n",
       "245          constitutive SLA-DR and SLA-DQ expression  \n",
       "82               murine and human p40 promoter regions  \n",
       "333                                genes UL49 and UL50  \n",
       "91          Sp1- and Ets-related transcription factors  \n",
       "72                         interferon-alpha and -gamma  \n",
       "241  both B-cell-restricted and ubiquitous transcri...  \n",
       "208       M-TAT/EPO, M-TAT/GM-CSF, and M-TAT/SCF cells  \n",
       "110                 kappa and lambda light-chain genes  \n",
       "209       M-TAT/EPO, M-TAT/GM-CSF, and M-TAT/SCF cells  \n",
       "3                                         c- and v-rel  \n",
       "297           erythroid differentiation and maturation  \n",
       "128                    c-fos and c-jun mRNA expression  \n",
       "393                           murine Th1 and Th2 cells  \n",
       "413                 Recombinant murine and human TCF-1  \n",
       "64                                    TRAF1, -2, or -3  \n",
       "140               macrophage and epithelial cell lines  \n",
       "65                                    TRAF1, -2, or -3  \n",
       "356                      PKC isoenzymes alpha and beta  \n",
       "71                           B- and T-cell development  \n",
       "122  abnormal cAMP and cAMP-phosphodiesterase activity  \n",
       "213  the erythroid, megakaryocytic, and mastocytic/...  \n",
       "166                    cyclin-dependent-kinase-4 and 6  \n",
       "212  the erythroid, megakaryocytic, and mastocytic/...  \n",
       "451                     NF-AT, AP-1 and CLE0 sequences  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['col_skip'].sort_values('count', ascending=False).head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for latex table\n",
    "import re\n",
    "csv_string = dfs['full'].sort_values('count', ascending=False).head(15).to_csv(sep='&')\n",
    "csv_string = re.sub(r'\\n\\d+&', r'  \\\\\\\\ \\\\addlinespace\\n', csv_string)  # line breaks and row numbers\n",
    "csv_string = re.sub('^&', r'', csv_string)  # first &\n",
    "csv_string = re.sub('%', r'\\\\%', csv_string)  # \\% for latex\n",
    "csv_string = re.sub('Ø', r'\\\\O', csv_string)  # \\O for latex\n",
    "csv_string = re.sub('#([A-Za-z\\-]+)', r'\\\\textbf{\\1}', csv_string)  # bold typeface used words \n",
    "csv_string = re.sub('&', r'\\t&\\t', csv_string)  # more space on delimiters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "struct\t&\tpos-seq\t&\tcount\t&\t\\%\t&\texample  \\\\ \\addlinespace\n",
      "(NP (NP n (NP \\O)) c (NP n (NP \\O)) (NP n))\t&\tncnn\t&\t63\t&\t4.44\t&\talpha and beta subunits  \\\\ \\addlinespace\n",
      "(NP (NP (ADJP a) (NP \\O)) c (NP (ADJP a) (NP \\O)) (NP n))\t&\tacan\t&\t55\t&\t3.87\t&\tlymphoid and nonlymphoid cells  \\\\ \\addlinespace\n",
      "(NP t (NP n (NP \\O)) c (NP n (NP \\O)) (NP n))\t&\ttncnn\t&\t30\t&\t2.11\t&\tthe TSG101 and FHIT genes  \\\\ \\addlinespace\n",
      "(NP (NP (NP n) c (NP n)) n)\t&\tncnn\t&\t28\t&\t1.97\t&\tH9 and CEM cells  \\\\ \\addlinespace\n",
      "(NP (ADJP (ADJP a) c (ADJP a)) n)\t&\tacan\t&\t19\t&\t1.34\t&\tcell-cell and cell-matrix interactions  \\\\ \\addlinespace\n",
      "(NP (NP n (NP \\O)) c (NP n (NP \\O)) (NP n n))\t&\tncnnn\t&\t17\t&\t1.2\t&\tT and B cell co-cultures  \\\\ \\addlinespace\n",
      "(NP (ADJP (ADJP a) c (ADJP a)) n n)\t&\tacann\t&\t11\t&\t0.77\t&\tnormal and dex-resistant CEM cells  \\\\ \\addlinespace\n",
      "(NP t (NP (NP n) c (NP n)) n)\t&\ttncnn\t&\t11\t&\t0.77\t&\tthe GM-CSF and IL-2 genes  \\\\ \\addlinespace\n",
      "(NP n c n n)\t&\tncnn\t&\t10\t&\t0.7\t&\tB and T cells  \\\\ \\addlinespace\n",
      "(NP (NP n (ADJP \\O) (NP \\O)) c (NP f a (NP \\O)) (NP n))\t&\tncfan\t&\t9\t&\t0.63\t&\tTNF-alpha- or PMA-induced stimulation  \\\\ \\addlinespace\n",
      "(NP (NP (ADJP a) (NP \\O)) c (NP (ADJP a) (NP \\O)) (NP n n))\t&\tacann\t&\t9\t&\t0.63\t&\terythroid and megakaryocytic lineage cells  \\\\ \\addlinespace\n",
      "(NP t (ADJP (ADJP a) c (ADJP a)) n)\t&\ttacan\t&\t9\t&\t0.63\t&\tthe sensitive and resistant patients  \\\\ \\addlinespace\n",
      "(NP c (NP (ADJP a) (NP \\O)) c (NP (ADJP a) (NP \\O)) (NP n))\t&\tcacan\t&\t8\t&\t0.56\t&\tboth transcriptional and post-transcriptional mechanisms  \\\\ \\addlinespace\n",
      "(NP (NP n (NP \\O)) , (NP n (NP \\O)) , c (NP n (NP \\O)) (NP n))\t&\tn,n,cnn\t&\t8\t&\t0.56\t&\tM-TAT/EPO, M-TAT/GM-CSF, and M-TAT/SCF cells  \\\\ \\addlinespace\n",
      "(NP (NP c (NP n) c (NP n)) n)\t&\tcncnn\t&\t8\t&\t0.56\t&\tboth HTLV-I and HTLV-II Tax\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(csv_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Constituent('cell-cell and cell-matrix interactions'(553, 591))\\NP,\n",
       " Constituent('immune and inflammatory responses'(124, 157))\\NP,\n",
       " Constituent('steroid-sensitive and steroid-resistant asthma'(69, 115))\\NP,\n",
       " Constituent('humoral or cell-mediated immunity'(476, 509))\\NP,\n",
       " Constituent('monocytic and erythroid differentiation'(54, 93))\\NP,\n",
       " Constituent('erythroid and megakaryocytic phenotypes'(768, 807))\\NP,\n",
       " Constituent('control and infected cells'(535, 561))\\NP,\n",
       " Constituent('basal and Tax-mediated transcription'(851, 887))\\NP,\n",
       " Constituent('cellular and viral genes'(205, 229))\\NP,\n",
       " Constituent('immune and inflammatory responses'(303, 336))\\NP,\n",
       " Constituent('severe and symptomatic VKC'(309, 335))\\NP,\n",
       " Constituent('bronchial and alveolar cells'(786, 814))\\NP,\n",
       " Constituent('oxidant and antioxidant conditions'(1924, 1958))\\NP,\n",
       " Constituent('haematopoietic and solid tumours'(387, 419))\\NP,\n",
       " Constituent('immune and inflammatory responses'(353, 386))\\NP,\n",
       " Constituent('neonatal and adult cells'(1127, 1151))\\NP,\n",
       " Constituent('homodimeric or heterodimeric forms'(1351, 1385))\\NP,\n",
       " Constituent('young and aged subjects'(97, 120))\\NP,\n",
       " Constituent('lymphoid and myeloid cells'(1169, 1195))\\NP,\n",
       " Constituent('CD4+ or CD8+ cells'(748, 766))\\NP]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "collapsed_skip_structures['((#a c a) #n)']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can capture more examples under fewer labels if we only use POS-tag sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "CROSS_COUNTING = ['concept', 'super', 'skip']  # order: concept, super, skip\n",
    "\n",
    "MOD = '[an]'\n",
    "SUPER_SEQ_COLLAPSERS = {\n",
    "    # re.compile('n*(n,?)+cnn+'),\n",
    "    # re.compile('a*(a,?)+caa*n+')\n",
    "    #re.compile('[an]c[an]n'),  # base case: shared head\n",
    "    #re.compile('[an]ncn'),  # base case: shared modifier\n",
    "    #re.compile('([an],)+[an],?c[an]n'),  # exp case: enumeration of modifiers + shared head\n",
    "    #re.compile('[an]+(n,)+n,?cn'),  # exp case: enumeration of heads with shared modifier\n",
    "    #re.compile('[an]c[an][an]+n'),  # exp case: multi-word head\n",
    "    #re.compile('[an]+[an]ncn'),  # exp case: shared pre-modifiers\n",
    "    re.compile(f'({MOD}n?,?)+c{MOD}*n'),  # all\n",
    "}\n",
    "\n",
    "CONCEPT_SEQ_COLLAPSERS = {\n",
    "    re.compile('(a|n)+n')\n",
    "}\n",
    "SAMPLE_CATEGORIES = {'concept', 'super', 'skip', 'cross-count'}\n",
    "\n",
    "def make_sequence(tokens, collapser=None):\n",
    "    sequence = [t.mapped_pos() if not t.pos == '#' else '#'\n",
    "                for t in tokens]\n",
    "    sequence_str = ''.join(sequence)\n",
    "    if not collapser:\n",
    "        return sequence_str\n",
    "\n",
    "    for pos_pattern in collapser:\n",
    "        if pos_pattern.fullmatch(sequence_str):\n",
    "            return pos_pattern.pattern\n",
    "    return sequence_str  # sequence did not match a pattern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "disc_concepts = [dc for doc in corpus\n",
    "                 for dc in doc.get_annotations(anno.DiscontinuousConcept)\n",
    "                 if 'c' in {t.mapped_pos() for t in dc.get_spanned_tokens()}\n",
    "                 and len(dc.get_tokens()) < 10]\n",
    "\n",
    "samples_after_category = {}\n",
    "for ct in SAMPLE_CATEGORIES:\n",
    "    samples_after_category[ct] = defaultdict(list)\n",
    "\n",
    "for dc in disc_concepts:\n",
    "    concept_tokens = dc.get_tokens()\n",
    "    all_tokens = dc.get_spanned_tokens()\n",
    "    skip_tokens = [t if t in concept_tokens  # actual token\n",
    "                   else anno.Token(dc.document, (-1, -1), '#')  # skipped token\n",
    "                   for t in all_tokens]\n",
    "\n",
    "    cross_count_type = []\n",
    "    cross_count_example = []\n",
    "\n",
    "    concept_sequence = make_sequence(concept_tokens, CONCEPT_SEQ_COLLAPSERS)\n",
    "    concept_text = dc.get_covered_text()\n",
    "    samples_after_category['concept'][concept_sequence].append(dc)\n",
    "    if 'concept' in CROSS_COUNTING:\n",
    "        cross_count_type.append(concept_sequence)\n",
    "        cross_count_example.append(dc)\n",
    "\n",
    "    super_sequence = make_sequence(all_tokens, SUPER_SEQ_COLLAPSERS)\n",
    "    super_text = dc.get_spanned_text()\n",
    "    samples_after_category['super'][super_sequence].append(dc)\n",
    "    if 'super' in CROSS_COUNTING:\n",
    "        cross_count_type.append(super_sequence)\n",
    "        cross_count_example.append(dc)\n",
    "\n",
    "    skip_sequence = make_sequence(skip_tokens)\n",
    "    samples_after_category['skip'][skip_sequence].append(dc)\n",
    "    if 'skip' in CROSS_COUNTING:\n",
    "        cross_count_type.append(skip_sequence)\n",
    "        cross_count_example.append(dc)\n",
    "\n",
    "    samples_after_category['cross-count'][tuple(cross_count_type)].append(\n",
    "        tuple(cross_count_example)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dict = {'type': [], 'count': [], '%': [], 'example': []}\n",
    "for type_, items in samples_after_category['super'].items():\n",
    "    data_dict['type'].append(type_)\n",
    "    data_dict['count'].append(len(items))\n",
    "    data_dict['%'].append(round(len(items) / len(disc_concepts) * 100, 2))\n",
    "    data_dict['example'].append(items[0].get_spanned_text())\n",
    "    \n",
    "df = pd.DataFrame(data_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>count</th>\n",
       "      <th>%</th>\n",
       "      <th>example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>([an]n?,?)+c[an]*n</td>\n",
       "      <td>990</td>\n",
       "      <td>66.80</td>\n",
       "      <td>GM-CSF and IL-2 genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ncfan</td>\n",
       "      <td>45</td>\n",
       "      <td>3.04</td>\n",
       "      <td>AP-1- and NF-kappaB-binding sites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ncfann</td>\n",
       "      <td>16</td>\n",
       "      <td>1.08</td>\n",
       "      <td>Sp1- and Ets-related transcription factors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ndcd</td>\n",
       "      <td>15</td>\n",
       "      <td>1.01</td>\n",
       "      <td>Jak 1 and 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ncfn</td>\n",
       "      <td>11</td>\n",
       "      <td>0.74</td>\n",
       "      <td>c- and v-rel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ncfnn</td>\n",
       "      <td>8</td>\n",
       "      <td>0.54</td>\n",
       "      <td>B- and T-cell development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>vncnn</td>\n",
       "      <td>6</td>\n",
       "      <td>0.40</td>\n",
       "      <td>truncated TSG101 and FHIT transcripts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>and,d,d,d,cdn</td>\n",
       "      <td>5</td>\n",
       "      <td>0.34</td>\n",
       "      <td>molecular mass 40, 42, 70, 120, and 130 kDa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>fncnn</td>\n",
       "      <td>5</td>\n",
       "      <td>0.34</td>\n",
       "      <td>HLA-DR or -DP transcription</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>fncd</td>\n",
       "      <td>5</td>\n",
       "      <td>0.34</td>\n",
       "      <td>cyclin-dependent-kinase-4 and 6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>fncn</td>\n",
       "      <td>5</td>\n",
       "      <td>0.34</td>\n",
       "      <td>interferon-alpha and -gamma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>acfn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>ER+ and ER- cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>facann</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>NF-kappaB-dependent and -independent signaling...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>aafn,n,nfn,cnn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>recombinant human nm23-H1, -H2, mouse nm23-M1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>nn,ncncrndn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>Categories Va, VI and DFR but not Category III...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>ndcndn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>type I and type II receptors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>nfacan</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>NF-kappa B-dependent and -independent pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>vacann</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>cultured human and murine cell lines</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>ancfnn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>human T- and B-cell lines</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>nndcndn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>Corticosteroid Type I and Type II receptors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>nndcnn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>protein phosphatase 1 and 2A inhibitors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>facan</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>PKC- dependent and -independent activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>nand,d,d,cdn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>EBV nuclear proteins 2, 3, 4, and 6 genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ncfaan</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>gC2- and gD2-specific CD4+ clones</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>nd,d,cd</td>\n",
       "      <td>4</td>\n",
       "      <td>0.27</td>\n",
       "      <td>sites 1, 2, and 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>vcan</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>transformed and normal lymphocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>facfan</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>eu -and hyperthyroid patients</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ncn(n)n</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>c-rel and v-rel (proto)oncogenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>n,n,cfan</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>STAT1-, STAT3-, and STAT5-related proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>n,cfan</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>STAT3-, and STAT5-related proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>annndcd</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>major histocompatibility complex classes I and II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ncnfan</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>AP-1- and NF-kappaB factor-binding elements</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>acfnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>gluco- and mineralocorticoid receptors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>ndcnnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>alpha 2 and gamma globin genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>n,cfann</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>PMA-, and RA-inducible CD23 expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>fn,d,cd</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>TRAF1, -2, or -3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>acvn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>quiescent and stimulated cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>nd,d,d,cd</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>positions -30, -44, -213, and -579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>acfan</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20</td>\n",
       "      <td>anti-inflammatory and anti-allergic/hypereosin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>nncnfan</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>cyclin E-Cdk2- and cyclin A-Cdk2-dependent pho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>ncfannn</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>p50- or p65- selected kappa B motif</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>nd,d,d,d,cd</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>Residues 432, 435, 437, 438, and 440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>nndcd</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>collagens type I and IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>fncdcfncdnn</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>GATA-1 and -2 or GATA-1 and -3 gene products</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>fncdnn</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>GATA-1 and -3 gene products</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>ann(n)ncnnn</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>basic leucine zipper (bZIP) DNA-binding and pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>ancan(n)n</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>possible T-cell and natural killer (T/NK) popu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>ncnfann</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>HIV-1- and interleukin 2-directed gene expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>aanfa(a)ca(a)n</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>clonogenetic fetal calf serum-supplemented (FC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>annnndcndnn</td>\n",
       "      <td>2</td>\n",
       "      <td>0.13</td>\n",
       "      <td>human T-cell leukemia virus type I and type II...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   type  count      %  \\\n",
       "5    ([an]n?,?)+c[an]*n    990  66.80   \n",
       "0                 ncfan     45   3.04   \n",
       "30               ncfann     16   1.08   \n",
       "16                 ndcd     15   1.01   \n",
       "3                  ncfn     11   0.74   \n",
       "25                ncfnn      8   0.54   \n",
       "9                 vncnn      6   0.40   \n",
       "85        and,d,d,d,cdn      5   0.34   \n",
       "78                fncnn      5   0.34   \n",
       "57                 fncd      5   0.34   \n",
       "26                 fncn      5   0.34   \n",
       "148                acfn      4   0.27   \n",
       "115              facann      4   0.27   \n",
       "221      aafn,n,nfn,cnn      4   0.27   \n",
       "109         nn,ncncrndn      4   0.27   \n",
       "106              ndcndn      4   0.27   \n",
       "29               nfacan      4   0.27   \n",
       "204              vacann      4   0.27   \n",
       "128              ancfnn      4   0.27   \n",
       "70              nndcndn      4   0.27   \n",
       "223              nndcnn      4   0.27   \n",
       "75                facan      4   0.27   \n",
       "247        nand,d,d,cdn      4   0.27   \n",
       "13               ncfaan      4   0.27   \n",
       "130             nd,d,cd      4   0.27   \n",
       "97                 vcan      3   0.20   \n",
       "36               facfan      3   0.20   \n",
       "2               ncn(n)n      3   0.20   \n",
       "138            n,n,cfan      3   0.20   \n",
       "139              n,cfan      3   0.20   \n",
       "143             annndcd      3   0.20   \n",
       "1                ncnfan      3   0.20   \n",
       "125               acfnn      3   0.20   \n",
       "226              ndcnnn      3   0.20   \n",
       "38              n,cfann      3   0.20   \n",
       "21              fn,d,cd      3   0.20   \n",
       "80                 acvn      3   0.20   \n",
       "31            nd,d,d,cd      3   0.20   \n",
       "92                acfan      3   0.20   \n",
       "87              nncnfan      2   0.13   \n",
       "132             ncfannn      2   0.13   \n",
       "104         nd,d,d,d,cd      2   0.13   \n",
       "93                nndcd      2   0.13   \n",
       "100         fncdcfncdnn      2   0.13   \n",
       "101              fncdnn      2   0.13   \n",
       "127         ann(n)ncnnn      2   0.13   \n",
       "111           ancan(n)n      2   0.13   \n",
       "90              ncnfann      2   0.13   \n",
       "122      aanfa(a)ca(a)n      2   0.13   \n",
       "89          annnndcndnn      2   0.13   \n",
       "\n",
       "                                               example  \n",
       "5                                GM-CSF and IL-2 genes  \n",
       "0                    AP-1- and NF-kappaB-binding sites  \n",
       "30          Sp1- and Ets-related transcription factors  \n",
       "16                                         Jak 1 and 3  \n",
       "3                                         c- and v-rel  \n",
       "25                           B- and T-cell development  \n",
       "9                truncated TSG101 and FHIT transcripts  \n",
       "85         molecular mass 40, 42, 70, 120, and 130 kDa  \n",
       "78                         HLA-DR or -DP transcription  \n",
       "57                     cyclin-dependent-kinase-4 and 6  \n",
       "26                         interferon-alpha and -gamma  \n",
       "148                                  ER+ and ER- cells  \n",
       "115  NF-kappaB-dependent and -independent signaling...  \n",
       "221  recombinant human nm23-H1, -H2, mouse nm23-M1,...  \n",
       "109  Categories Va, VI and DFR but not Category III...  \n",
       "106                       type I and type II receptors  \n",
       "29      NF-kappa B-dependent and -independent pathways  \n",
       "204               cultured human and murine cell lines  \n",
       "128                          human T- and B-cell lines  \n",
       "70         Corticosteroid Type I and Type II receptors  \n",
       "223            protein phosphatase 1 and 2A inhibitors  \n",
       "75          PKC- dependent and -independent activation  \n",
       "247          EBV nuclear proteins 2, 3, 4, and 6 genes  \n",
       "13                   gC2- and gD2-specific CD4+ clones  \n",
       "130                                  sites 1, 2, and 3  \n",
       "97                  transformed and normal lymphocytes  \n",
       "36                       eu -and hyperthyroid patients  \n",
       "2                     c-rel and v-rel (proto)oncogenes  \n",
       "138         STAT1-, STAT3-, and STAT5-related proteins  \n",
       "139                 STAT3-, and STAT5-related proteins  \n",
       "143  major histocompatibility complex classes I and II  \n",
       "1          AP-1- and NF-kappaB factor-binding elements  \n",
       "125             gluco- and mineralocorticoid receptors  \n",
       "226                     alpha 2 and gamma globin genes  \n",
       "38              PMA-, and RA-inducible CD23 expression  \n",
       "21                                    TRAF1, -2, or -3  \n",
       "80                      quiescent and stimulated cells  \n",
       "31                  positions -30, -44, -213, and -579  \n",
       "92   anti-inflammatory and anti-allergic/hypereosin...  \n",
       "87   cyclin E-Cdk2- and cyclin A-Cdk2-dependent pho...  \n",
       "132                p50- or p65- selected kappa B motif  \n",
       "104               Residues 432, 435, 437, 438, and 440  \n",
       "93                             collagens type I and IV  \n",
       "100       GATA-1 and -2 or GATA-1 and -3 gene products  \n",
       "101                        GATA-1 and -3 gene products  \n",
       "127  basic leucine zipper (bZIP) DNA-binding and pr...  \n",
       "111  possible T-cell and natural killer (T/NK) popu...  \n",
       "90   HIV-1- and interleukin 2-directed gene expression  \n",
       "122  clonogenetic fetal calf serum-supplemented (FC...  \n",
       "89   human T-cell leukemia virus type I and type II...  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_values('count', ascending=False).head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IL-4 receptor subunits alpha and gamma\n",
      "viral gene expression and replication\n",
      "viral transcription initiation and elongation\n",
      "HTLV-I basal transcription and expression\n",
      "B cell proliferation and differentiation\n",
      "PKC isoenzymes alpha and beta\n",
      "T cell receptor alpha and -beta\n",
      "human glial cells and lymphocytes\n",
      "human B cell proliferation and differentiation\n",
      "human GM-CSF promoter and enhancer\n",
      "T lymphocyte differentiation and activation\n",
      "human chromosomes 11p15 and 11p13\n",
      "T lymphocyte activation and mitogenesis\n",
      "T cell activation and differentiation\n",
      "T cell activation and growth\n"
     ]
    }
   ],
   "source": [
    "for item in samples_after_category['super']['[an]+[an]ncn']:\n",
    "    print(*(t.get_covered_text() for t in item.get_spanned_tokens()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>struct</th>\n",
       "      <th>pos-seq</th>\n",
       "      <th>count</th>\n",
       "      <th>%</th>\n",
       "      <th>example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>(a n n c n)</td>\n",
       "      <td>anncn</td>\n",
       "      <td>2</td>\n",
       "      <td>0.14</td>\n",
       "      <td>viral gene expression and replication</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>(n ((a n) c n))</td>\n",
       "      <td>nancn</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>HTLV-I basal transcription and expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>(n n n c n)</td>\n",
       "      <td>nnncn</td>\n",
       "      <td>2</td>\n",
       "      <td>0.14</td>\n",
       "      <td>B cell proliferation and differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>(n n (n c n))</td>\n",
       "      <td>nnncn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>PKC isoenzymes alpha and beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>(n n n n c n)</td>\n",
       "      <td>nnnncn</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>T cell receptor alpha and -beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>(a (a n) c n)</td>\n",
       "      <td>aancn</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>human glial cells and lymphocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>(a n n (n c n))</td>\n",
       "      <td>annncn</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>human B cell proliferation and differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>(a n (n c n))</td>\n",
       "      <td>anncn</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>human chromosomes 11p15 and 11p13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              struct pos-seq  count     %  \\\n",
       "101      (a n n c n)   anncn      2  0.14   \n",
       "218  (n ((a n) c n))   nancn      1  0.07   \n",
       "262      (n n n c n)   nnncn      2  0.14   \n",
       "275    (n n (n c n))   nnncn      3  0.21   \n",
       "321    (n n n n c n)  nnnncn      1  0.07   \n",
       "345    (a (a n) c n)   aancn      1  0.07   \n",
       "389  (a n n (n c n))  annncn      1  0.07   \n",
       "528    (a n (n c n))   anncn      1  0.07   \n",
       "\n",
       "                                            example  \n",
       "101           viral gene expression and replication  \n",
       "218       HTLV-I basal transcription and expression  \n",
       "262        B cell proliferation and differentiation  \n",
       "275                   PKC isoenzymes alpha and beta  \n",
       "321                 T cell receptor alpha and -beta  \n",
       "345               human glial cells and lymphocytes  \n",
       "389  human B cell proliferation and differentiation  \n",
       "528               human chromosomes 11p15 and 11p13  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for exp case: multi-word head: for multiple tokens before/after the conjunction,\n",
    "# do the tokens go with the head or the second modifier?\n",
    "df = dfs['col_full']\n",
    "subset = df[df['pos-seq'].str.contains('^[an]+[an]ncn$')]\n",
    "subset.head(25)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What about CC's?'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_concepts = [c for doc in corpus for c in doc.get_annotations(anno.Concept)\n",
    "               if not isinstance(c, anno.DiscontinuousConcept) and not len(c) < 2]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "dist_of_types = Counter(c.pos_sequence() for c in all_concepts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('nn', 15031),\n",
       " ('an', 7388),\n",
       " ('nnn', 4569),\n",
       " ('ann', 2870),\n",
       " ('aan', 1206),\n",
       " ('nnnn', 1006),\n",
       " ('fan', 865),\n",
       " ('annn', 831),\n",
       " ('nan', 617),\n",
       " ('nd', 384)]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dist_of_types.most_common(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "simple_count = sum(v for t, v in dist_of_types.items() if re.match('[an]n', t))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7885333579233713"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simple_count / sum(dist_of_types.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
