{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Structural analysis of discontinuous concepts\n",
    "\n",
    "The purpose of this analysis notebook is to make an in-depth analysis of structures. This makes the basis for a large portion of the corresponding section in the final thesis paper."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Coordinated concepts\n",
    "\n",
    "Coordinations are the only thing marked in GENIA while CRAFT also contains other types. Therefore, GENIA will be used primarily in this part. The variable `corpus_name` can be set to either `genia` or `craft` to run with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "corpus_name = 'genia'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, let's load the corpus, then extract concepts and their constituent structures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading GENIA corpus ... NOTE: 13 files cannot get Constituent annotations!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1599/1599 [00:19<00:00, 83.40it/s] \n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "os.chdir('..')  # get to the root directory of the project\n",
    "\n",
    "from datautils import dataio, annotations as anno\n",
    "\n",
    "# load the corpora\n",
    "corpus = dataio.load_corpus(corpus_name)\n",
    "\n",
    "# not all docs in genia get Constituent annotations; if so, leave them out\n",
    "if corpus_name.lower() == 'genia':\n",
    "    corpus = [doc for doc in corpus if doc.get_annotations(anno.Constituent)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "concept_consts = [(concept, doc.get_annotations_at(concept.span, anno.Constituent)[0])\n",
    "                  for doc in corpus\n",
    "                  for concept in doc.get_annotations(anno.DiscontinuousConcept)\n",
    "                  if len(concept.get_tokens()) < 10]\n",
    "\n",
    "# for craft, keep only the ones with a conjunction in them\n",
    "temp_list = []\n",
    "for concept, const in concept_consts:\n",
    "    present_pos = {t.mapped_pos() for t in const.get_tokens()}\n",
    "    if 'c' in present_pos:\n",
    "        temp_list.append((concept, const))\n",
    "concept_consts = temp_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each constituent can spit out a treebank structure. Let's do that and put it all in dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def skipped_structure(const, allowed_tokens):\n",
    "    \"\"\"Returns  a treebank structure of a Constituent with a # symbol in front of\n",
    "    all tokens that are in allowed_tokens\"\"\"\n",
    "    # note: craft constituents often have labels like NP-SBJ; keep only the first part\n",
    "    return '(' + const.label.split('-')[0] + ' ' + ' '.join(\n",
    "        skipped_structure(c, allowed_tokens) if isinstance(c, anno.Constituent)\n",
    "        else c.mapped_pos() if c not in allowed_tokens\n",
    "        else '#' + c.mapped_pos()\n",
    "        for c in const.constituents\n",
    "    ) + ')'\n",
    "\n",
    "def unlabeled_structure(const, allowed_tokens):\n",
    "    \"\"\"Returns a simplified treebank structure of a Constituent with no constituent labels\n",
    "    and with a # symbol in front of all tokens that are in allowed_tokens\"\"\"\n",
    "    return_string = '(' + ' '.join(\n",
    "        unlabeled_structure(c, allowed_tokens) if isinstance(c, anno.Constituent)\n",
    "        else c.mapped_pos() if c not in allowed_tokens\n",
    "        else '#' + c.mapped_pos()\n",
    "        for c in const.constituents\n",
    "        if not contains_empty_token(c)\n",
    "    ) + ')'\n",
    "    # remove all cases of single token constituents, e.g. (n)\n",
    "    return re.sub('\\(+(#?.)\\)+', r'\\1', return_string)\n",
    "\n",
    "def contains_empty_token(const):\n",
    "    if not isinstance(const, anno.Constituent):\n",
    "        return False\n",
    "    for c in const.constituents:\n",
    "        if isinstance(c, anno.Token) and c.mapped_pos() == 'Ø':\n",
    "            return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "\n",
    "full_structures = defaultdict(list)\n",
    "skip_structures = defaultdict(list)\n",
    "collapsed_full_structures = defaultdict(list)\n",
    "collapsed_skip_structures = defaultdict(list)\n",
    "\n",
    "for concept, const in concept_consts:\n",
    "    struct = const.structure()  \n",
    "    full_structures[struct].append(const)\n",
    "    \n",
    "    skip_struct = skipped_structure(const, concept.get_concept_tokens())\n",
    "    skip_structures[skip_struct].append(const)\n",
    "    \n",
    "    collapsed_full = unlabeled_structure(const, {})\n",
    "    collapsed_full_structures[collapsed_full].append(const)\n",
    "    \n",
    "    collapsed_skip = unlabeled_structure(const, concept.get_concept_tokens())\n",
    "    collapsed_skip_structures[collapsed_skip].append(const)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_dcs = len(concept_consts)\n",
    "\n",
    "dfs = {}\n",
    "\n",
    "for name, structure_dict in [('full', full_structures),\n",
    "                             ('skip', skip_structures),\n",
    "                             ('col_full', collapsed_full_structures),\n",
    "                             ('col_skip', collapsed_skip_structures)\n",
    "                            ]:\n",
    "    \n",
    "    data_dict = {'struct': [], 'pos-seq': [], 'count': [], '%': [], 'example': []}\n",
    "    \n",
    "    for struct, sample in structure_dict.items():\n",
    "        data_dict['struct'].append(struct)\n",
    "        data_dict['pos-seq'].append(''.join(c for c in struct if c.islower() or c in ','))\n",
    "        data_dict['count'].append(len(sample))\n",
    "        data_dict['%'].append(round(len(sample) / n_dcs * 100, 2))\n",
    "        data_dict['example'].append(sample[0].get_covered_text())\n",
    "        \n",
    "    dfs[name] = pd.DataFrame(data_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>struct</th>\n",
       "      <th>pos-seq</th>\n",
       "      <th>count</th>\n",
       "      <th>%</th>\n",
       "      <th>example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>(#n c n #n)</td>\n",
       "      <td>ncnn</td>\n",
       "      <td>83</td>\n",
       "      <td>5.84</td>\n",
       "      <td>T and NK cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>(#a c a #n)</td>\n",
       "      <td>acan</td>\n",
       "      <td>69</td>\n",
       "      <td>4.86</td>\n",
       "      <td>Positive and negative regulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(t #n c n #n)</td>\n",
       "      <td>tncnn</td>\n",
       "      <td>44</td>\n",
       "      <td>3.10</td>\n",
       "      <td>the FY*A and FY*B alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>((#n c n) #n)</td>\n",
       "      <td>ncnn</td>\n",
       "      <td>33</td>\n",
       "      <td>2.32</td>\n",
       "      <td>Egr-1 and c-fos mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>((#a c a) #n)</td>\n",
       "      <td>acan</td>\n",
       "      <td>21</td>\n",
       "      <td>1.48</td>\n",
       "      <td>control and infected cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>(#n c n (#n #n))</td>\n",
       "      <td>ncnnn</td>\n",
       "      <td>20</td>\n",
       "      <td>1.41</td>\n",
       "      <td>HLA-DMA and HLA-DMB gene expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>(#n c (f #a) #n)</td>\n",
       "      <td>ncfan</td>\n",
       "      <td>14</td>\n",
       "      <td>0.99</td>\n",
       "      <td>TNF-alpha- or PMA-induced stimulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>(#a c a (#n #n))</td>\n",
       "      <td>acann</td>\n",
       "      <td>14</td>\n",
       "      <td>0.99</td>\n",
       "      <td>bacterial or mycobacterial response elements</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>((#a c a) #n #n)</td>\n",
       "      <td>acann</td>\n",
       "      <td>12</td>\n",
       "      <td>0.84</td>\n",
       "      <td>endocrine and exocrine precursor cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>((#n c (f #a)) #n)</td>\n",
       "      <td>ncfan</td>\n",
       "      <td>12</td>\n",
       "      <td>0.84</td>\n",
       "      <td>glucocorticoid- and Fas-mediated cytotoxicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>(t (#n c n) #n)</td>\n",
       "      <td>tncnn</td>\n",
       "      <td>12</td>\n",
       "      <td>0.84</td>\n",
       "      <td>the AP-1 and Elf-1 sites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>(t #a c a #n)</td>\n",
       "      <td>tacan</td>\n",
       "      <td>11</td>\n",
       "      <td>0.77</td>\n",
       "      <td>the extracellular and transmembrane domains</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>(t (#a c a) #n)</td>\n",
       "      <td>tacan</td>\n",
       "      <td>11</td>\n",
       "      <td>0.77</td>\n",
       "      <td>a cytoplasmic and/or nuclear receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>(c #a c a #n)</td>\n",
       "      <td>cacan</td>\n",
       "      <td>11</td>\n",
       "      <td>0.77</td>\n",
       "      <td>both transcriptional and posttranscriptional l...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>((#n c n) #n #n)</td>\n",
       "      <td>ncnnn</td>\n",
       "      <td>10</td>\n",
       "      <td>0.70</td>\n",
       "      <td>Egr-1 and c-fos mRNA expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>((c #n c n) #n)</td>\n",
       "      <td>cncnn</td>\n",
       "      <td>8</td>\n",
       "      <td>0.56</td>\n",
       "      <td>both M-CSF and c-fms expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550</th>\n",
       "      <td>(#a c n #n)</td>\n",
       "      <td>acnn</td>\n",
       "      <td>8</td>\n",
       "      <td>0.56</td>\n",
       "      <td>myeloid and B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>((#a c a) #a #n)</td>\n",
       "      <td>acaan</td>\n",
       "      <td>7</td>\n",
       "      <td>0.49</td>\n",
       "      <td>erythroid and non-erythroid nuclear proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>(#n (d c #d))</td>\n",
       "      <td>ndcd</td>\n",
       "      <td>7</td>\n",
       "      <td>0.49</td>\n",
       "      <td>positions -585 and -650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>((c #a c a) #n)</td>\n",
       "      <td>cacan</td>\n",
       "      <td>7</td>\n",
       "      <td>0.49</td>\n",
       "      <td>both CD4(+) and CD8(+) subsets</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>640</th>\n",
       "      <td>(#n c n (#n #n #n))</td>\n",
       "      <td>ncnnnn</td>\n",
       "      <td>7</td>\n",
       "      <td>0.49</td>\n",
       "      <td>Sp1 and CREB binding site oligonucleotides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>(#n c n #n #n)</td>\n",
       "      <td>ncnnn</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>B and T lymphocyte activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>(t #n c n (#n #n))</td>\n",
       "      <td>tncnnn</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>The GATA-1 and Sp1 binding sites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(#n c (f #n))</td>\n",
       "      <td>ncfn</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>sIgM- and sIgD-crosslinking</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>(a #n c n #n)</td>\n",
       "      <td>ancnn</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>multiple Ets and NF-kappaB/NFAT proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>(#n c (f #a) (#n #n))</td>\n",
       "      <td>ncfann</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>androgen- and antiandrogen-bound receptor comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>(#n d c #d)</td>\n",
       "      <td>ndcd</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42</td>\n",
       "      <td>Helices I and II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>(a #a c a #n)</td>\n",
       "      <td>aacan</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>many viral and cellular promoters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>((#f n) c #n)</td>\n",
       "      <td>fncn</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>IFN-alpha and -beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>(c #n c n #n)</td>\n",
       "      <td>cncnn</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>both IL-12 and B7.1 cDNAs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>(#n , n , c n #n)</td>\n",
       "      <td>n,n,cnn</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>Th0, Th1, and Th2 clones</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>(#n n c #n)</td>\n",
       "      <td>nncn</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>DP alpha and -beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>(n , #n , c n #n)</td>\n",
       "      <td>n,n,cnn</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>Th0, Th1, and Th2 clones</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>((#f n) c #d)</td>\n",
       "      <td>fncd</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>cyclin-dependent-kinase-4 and 6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>(#n c n (#a #n))</td>\n",
       "      <td>ncnan</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35</td>\n",
       "      <td>Stat1 and Stat3 transcriptional activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>(#a n c #n)</td>\n",
       "      <td>ancn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>homotypic aggregation and proliferation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>(#a c a (#n #n #n))</td>\n",
       "      <td>acannn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>human and murine T-ALL cell lines</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>(c #a c a (#n #n))</td>\n",
       "      <td>cacann</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>either 5' or 3' NF-kappaB sites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>(#a #n c n #n)</td>\n",
       "      <td>ancnn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>Human T and B lymphocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>(#n (n c #n))</td>\n",
       "      <td>nncn</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>B-cell growth and development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>((#n c (f #a)) #n #n)</td>\n",
       "      <td>ncfann</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28</td>\n",
       "      <td>TNF-alpha- and PMA-stimulated U1 cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>((#f n) , #d , c d)</td>\n",
       "      <td>fn,d,cd</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>IL-2, -7, and -15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>(t (n , #n , c n) #n)</td>\n",
       "      <td>tn,n,cnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>the muA, muE3, and muB sites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>(c t #n c n #n)</td>\n",
       "      <td>ctncnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>both the M-CSF and FMS genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>(t (#n c (f #a)) #n)</td>\n",
       "      <td>tncfan</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>a TNF-alpha- and NF-kappaB-dependent mechanism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>(#n #n (n c #n))</td>\n",
       "      <td>nnncn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>PKC isoenzymes alpha and beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>((#n #n) c n #n)</td>\n",
       "      <td>nncnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>IL-2 mRNA and protein levels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>(#a c n (#n #n))</td>\n",
       "      <td>acnnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>human and mouse TCF-1 protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>((n , #n c n) #n)</td>\n",
       "      <td>n,ncnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>STAT6, STAT3 and STAT5 proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>((#n , n c n) #n)</td>\n",
       "      <td>n,ncnn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>STAT6, STAT3 and STAT5 proteins</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    struct   pos-seq  count     %  \\\n",
       "13             (#n c n #n)      ncnn     83  5.84   \n",
       "9              (#a c a #n)      acan     69  4.86   \n",
       "0            (t #n c n #n)     tncnn     44  3.10   \n",
       "6            ((#n c n) #n)      ncnn     33  2.32   \n",
       "50           ((#a c a) #n)      acan     21  1.48   \n",
       "15        (#n c n (#n #n))     ncnnn     20  1.41   \n",
       "36        (#n c (f #a) #n)     ncfan     14  0.99   \n",
       "29        (#a c a (#n #n))     acann     14  0.99   \n",
       "107       ((#a c a) #n #n)     acann     12  0.84   \n",
       "215     ((#n c (f #a)) #n)     ncfan     12  0.84   \n",
       "49         (t (#n c n) #n)     tncnn     12  0.84   \n",
       "166          (t #a c a #n)     tacan     11  0.77   \n",
       "96         (t (#a c a) #n)     tacan     11  0.77   \n",
       "27           (c #a c a #n)     cacan     11  0.77   \n",
       "3         ((#n c n) #n #n)     ncnnn     10  0.70   \n",
       "101        ((c #n c n) #n)     cncnn      8  0.56   \n",
       "550            (#a c n #n)      acnn      8  0.56   \n",
       "423       ((#a c a) #a #n)     acaan      7  0.49   \n",
       "46           (#n (d c #d))      ndcd      7  0.49   \n",
       "83         ((c #a c a) #n)     cacan      7  0.49   \n",
       "640    (#n c n (#n #n #n))    ncnnnn      7  0.49   \n",
       "179         (#n c n #n #n)     ncnnn      6  0.42   \n",
       "60      (t #n c n (#n #n))    tncnnn      6  0.42   \n",
       "4            (#n c (f #n))      ncfn      6  0.42   \n",
       "193          (a #n c n #n)     ancnn      6  0.42   \n",
       "306  (#n c (f #a) (#n #n))    ncfann      6  0.42   \n",
       "167            (#n d c #d)      ndcd      6  0.42   \n",
       "66           (a #a c a #n)     aacan      5  0.35   \n",
       "100          ((#f n) c #n)      fncn      5  0.35   \n",
       "32           (c #n c n #n)     cncnn      5  0.35   \n",
       "98       (#n , n , c n #n)   n,n,cnn      5  0.35   \n",
       "17             (#n n c #n)      nncn      5  0.35   \n",
       "99       (n , #n , c n #n)   n,n,cnn      5  0.35   \n",
       "194          ((#f n) c #d)      fncd      5  0.35   \n",
       "140       (#n c n (#a #n))     ncnan      5  0.35   \n",
       "141            (#a n c #n)      ancn      4  0.28   \n",
       "234    (#a c a (#n #n #n))    acannn      4  0.28   \n",
       "97      (c #a c a (#n #n))    cacann      4  0.28   \n",
       "180         (#a #n c n #n)     ancnn      4  0.28   \n",
       "500          (#n (n c #n))      nncn      4  0.28   \n",
       "441  ((#n c (f #a)) #n #n)    ncfann      4  0.28   \n",
       "368    ((#f n) , #d , c d)   fn,d,cd      3  0.21   \n",
       "273  (t (n , #n , c n) #n)  tn,n,cnn      3  0.21   \n",
       "256        (c t #n c n #n)    ctncnn      3  0.21   \n",
       "113   (t (#n c (f #a)) #n)    tncfan      3  0.21   \n",
       "115       (#n #n (n c #n))     nnncn      3  0.21   \n",
       "86        ((#n #n) c n #n)     nncnn      3  0.21   \n",
       "378       (#a c n (#n #n))     acnnn      3  0.21   \n",
       "85       ((n , #n c n) #n)    n,ncnn      3  0.21   \n",
       "84       ((#n , n c n) #n)    n,ncnn      3  0.21   \n",
       "\n",
       "                                               example  \n",
       "13                                      T and NK cells  \n",
       "9                     Positive and negative regulation  \n",
       "0                            the FY*A and FY*B alleles  \n",
       "6                                 Egr-1 and c-fos mRNA  \n",
       "50                          control and infected cells  \n",
       "15                 HLA-DMA and HLA-DMB gene expression  \n",
       "36               TNF-alpha- or PMA-induced stimulation  \n",
       "29        bacterial or mycobacterial response elements  \n",
       "107             endocrine and exocrine precursor cells  \n",
       "215      glucocorticoid- and Fas-mediated cytotoxicity  \n",
       "49                            the AP-1 and Elf-1 sites  \n",
       "166        the extracellular and transmembrane domains  \n",
       "96               a cytoplasmic and/or nuclear receptor  \n",
       "27   both transcriptional and posttranscriptional l...  \n",
       "3                      Egr-1 and c-fos mRNA expression  \n",
       "101                    both M-CSF and c-fms expression  \n",
       "550                                myeloid and B cells  \n",
       "423       erythroid and non-erythroid nuclear proteins  \n",
       "46                             positions -585 and -650  \n",
       "83                      both CD4(+) and CD8(+) subsets  \n",
       "640         Sp1 and CREB binding site oligonucleotides  \n",
       "179                      B and T lymphocyte activation  \n",
       "60                    The GATA-1 and Sp1 binding sites  \n",
       "4                          sIgM- and sIgD-crosslinking  \n",
       "193           multiple Ets and NF-kappaB/NFAT proteins  \n",
       "306  androgen- and antiandrogen-bound receptor comp...  \n",
       "167                                   Helices I and II  \n",
       "66                   many viral and cellular promoters  \n",
       "100                                IFN-alpha and -beta  \n",
       "32                           both IL-12 and B7.1 cDNAs  \n",
       "98                            Th0, Th1, and Th2 clones  \n",
       "17                                  DP alpha and -beta  \n",
       "99                            Th0, Th1, and Th2 clones  \n",
       "194                    cyclin-dependent-kinase-4 and 6  \n",
       "140           Stat1 and Stat3 transcriptional activity  \n",
       "141            homotypic aggregation and proliferation  \n",
       "234                  human and murine T-ALL cell lines  \n",
       "97                     either 5' or 3' NF-kappaB sites  \n",
       "180                          Human T and B lymphocytes  \n",
       "500                      B-cell growth and development  \n",
       "441             TNF-alpha- and PMA-stimulated U1 cells  \n",
       "368                                  IL-2, -7, and -15  \n",
       "273                       the muA, muE3, and muB sites  \n",
       "256                       both the M-CSF and FMS genes  \n",
       "113     a TNF-alpha- and NF-kappaB-dependent mechanism  \n",
       "115                      PKC isoenzymes alpha and beta  \n",
       "86                        IL-2 mRNA and protein levels  \n",
       "378                      human and mouse TCF-1 protein  \n",
       "85                     STAT6, STAT3 and STAT5 proteins  \n",
       "84                     STAT6, STAT3 and STAT5 proteins  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['col_skip'].sort_values('count', ascending=False).head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for latex table\n",
    "import re\n",
    "csv_string = dfs['full'].sort_values('count', ascending=False).head(15).to_csv(sep='&')\n",
    "csv_string = re.sub(r'\\n\\d+&', r'  \\\\\\\\ \\\\addlinespace\\n', csv_string)  # line breaks and row numbers\n",
    "csv_string = re.sub('^&', r'', csv_string)  # first &\n",
    "csv_string = re.sub('%', r'\\\\%', csv_string)  # \\% for latex\n",
    "csv_string = re.sub('Ø', r'\\\\O', csv_string)  # \\O for latex\n",
    "csv_string = re.sub('#([A-Za-z\\-]+)', r'\\\\textbf{\\1}', csv_string)  # bold typeface used words \n",
    "csv_string = re.sub('&', r'\\t&\\t', csv_string)  # more space on delimiters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "struct\t&\tpos-seq\t&\tcount\t&\t\\%\t&\texample  \\\\ \\addlinespace\n",
      "(NP (NP n (NP \\O)) c (NP n (NP \\O)) (NP n))\t&\tncnn\t&\t70\t&\t4.93\t&\tT and NK cells  \\\\ \\addlinespace\n",
      "(NP (NP (ADJP a) (NP \\O)) c (NP (ADJP a) (NP \\O)) (NP n))\t&\tacan\t&\t61\t&\t4.29\t&\tPositive and negative regulation  \\\\ \\addlinespace\n",
      "(NP t (NP n (NP \\O)) c (NP n (NP \\O)) (NP n))\t&\ttncnn\t&\t36\t&\t2.53\t&\tthe FY*A and FY*B alleles  \\\\ \\addlinespace\n",
      "(NP (NP (NP n) c (NP n)) n)\t&\tncnn\t&\t32\t&\t2.25\t&\tEgr-1 and c-fos mRNA  \\\\ \\addlinespace\n",
      "(NP (ADJP (ADJP a) c (ADJP a)) n)\t&\tacan\t&\t20\t&\t1.41\t&\tmonocytic and erythroid differentiation  \\\\ \\addlinespace\n",
      "(NP (NP n (NP \\O)) c (NP n (NP \\O)) (NP n n))\t&\tncnnn\t&\t20\t&\t1.41\t&\tHLA-DMA and HLA-DMB gene expression  \\\\ \\addlinespace\n",
      "(NP (NP n (ADJP \\O) (NP \\O)) c (NP f a (NP \\O)) (NP n))\t&\tncfan\t&\t13\t&\t0.91\t&\tTNF-alpha- or PMA-induced stimulation  \\\\ \\addlinespace\n",
      "(NP t (NP (NP n) c (NP n)) n)\t&\ttncnn\t&\t12\t&\t0.84\t&\tthe AP-1 and Elf-1 sites  \\\\ \\addlinespace\n",
      "(NP (ADJP (ADJP a) c (ADJP a)) n n)\t&\tacann\t&\t12\t&\t0.84\t&\tendocrine and exocrine precursor cells  \\\\ \\addlinespace\n",
      "(NP t (ADJP (ADJP a) c (ADJP a)) n)\t&\ttacan\t&\t11\t&\t0.77\t&\ta cytoplasmic and/or nuclear receptor  \\\\ \\addlinespace\n",
      "(NP (NP (ADJP a) (NP \\O)) c (NP (ADJP a) (NP \\O)) (NP n n))\t&\tacann\t&\t11\t&\t0.77\t&\tbacterial or mycobacterial response elements  \\\\ \\addlinespace\n",
      "(NP n c n n)\t&\tncnn\t&\t11\t&\t0.77\t&\tHL-60 and NB4 cells  \\\\ \\addlinespace\n",
      "(NP c (NP (ADJP a) (NP \\O)) c (NP (ADJP a) (NP \\O)) (NP n))\t&\tcacan\t&\t10\t&\t0.7\t&\tboth transcriptional and posttranscriptional levels  \\\\ \\addlinespace\n",
      "(NP (NP (NP n) c (NP n)) n n)\t&\tncnnn\t&\t10\t&\t0.7\t&\tEgr-1 and c-fos mRNA expression  \\\\ \\addlinespace\n",
      "(NP (NP n (NP \\O)) , (NP n (NP \\O)) , c (NP n (NP \\O)) (NP n))\t&\tn,n,cnn\t&\t10\t&\t0.7\t&\tTh0, Th1, and Th2 clones\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(csv_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Constituent('control and infected cells'(535, 561))\\NP,\n",
       " Constituent('monocytic and erythroid differentiation'(54, 93))\\NP,\n",
       " Constituent('bronchial and alveolar cells'(786, 814))\\NP,\n",
       " Constituent('steroid-sensitive and steroid-resistant asthma'(69, 115))\\NP,\n",
       " Constituent('severe and symptomatic VKC'(309, 335))\\NP,\n",
       " Constituent('humoral or cell-mediated immunity'(476, 509))\\NP,\n",
       " Constituent('immune and inflammatory responses'(303, 336))\\NP,\n",
       " Constituent('immune and inflammatory responses'(124, 157))\\NP,\n",
       " Constituent('haematopoietic and solid tumours'(387, 419))\\NP,\n",
       " Constituent('immune and inflammatory responses'(353, 386))\\NP,\n",
       " Constituent('neonatal and adult cells'(1127, 1151))\\NP,\n",
       " Constituent('basal and Tax-mediated transcription'(851, 887))\\NP,\n",
       " Constituent('oxidant and antioxidant conditions'(1924, 1958))\\NP,\n",
       " Constituent('Murine and human TCF-1'(703, 725))\\NP,\n",
       " Constituent('homodimeric or heterodimeric forms'(1351, 1385))\\NP,\n",
       " Constituent('cellular and viral genes'(205, 229))\\NP,\n",
       " Constituent('cell-cell and cell-matrix interactions'(553, 591))\\NP,\n",
       " Constituent('CD4+ or CD8+ cells'(748, 766))\\NP,\n",
       " Constituent('young and aged subjects'(97, 120))\\NP,\n",
       " Constituent('erythroid and megakaryocytic phenotypes'(768, 807))\\NP,\n",
       " Constituent('lymphoid and myeloid cells'(1169, 1195))\\NP]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "collapsed_skip_structures['((#a c a) #n)']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can capture more examples under fewer labels if we only use POS-tag sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "CROSS_COUNTING = ['concept', 'super', 'skip']  # order: concept, super, skip\n",
    "SUPER_SEQ_COLLAPSERS = {\n",
    "    # re.compile('n*(n,?)+cnn+'),\n",
    "    # re.compile('a*(a,?)+caa*n+')\n",
    "    re.compile('[an]c[an]n'),  # base case: shared head\n",
    "    re.compile('[an]ncn'),  # base case: shared modifier\n",
    "    re.compile('([an],)+[an],?c[an]n'),  # exp case: enumeration of modifiers + shared head\n",
    "    re.compile('[an]+(n,)+n,?cn'),  # exp case: enumeration of heads with shared modifier\n",
    "    re.compile('[an]c[an][an]+n'),  # exp case: multi-word head\n",
    "    re.compile('[an]+[an]ncn'),  # exp case: shared pre-modifiers\n",
    "}\n",
    "CONCEPT_SEQ_COLLAPSERS = {\n",
    "    re.compile('(a|n)+n')\n",
    "}\n",
    "SAMPLE_CATEGORIES = {'concept', 'super', 'skip', 'cross-count'}\n",
    "\n",
    "def make_sequence(tokens, collapser=None):\n",
    "    sequence = [t.mapped_pos() if not t.pos == '#' else '#'\n",
    "                for t in tokens]\n",
    "    sequence_str = ''.join(sequence)\n",
    "    if not collapser:\n",
    "        return sequence_str\n",
    "\n",
    "    for pos_pattern in collapser:\n",
    "        if pos_pattern.fullmatch(sequence_str):\n",
    "            return pos_pattern.pattern\n",
    "    return sequence_str  # sequence did not match a pattern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "disc_concepts = [dc for doc in corpus\n",
    "                 for dc in doc.get_annotations(anno.DiscontinuousConcept)\n",
    "                 if 'c' in {t.mapped_pos() for t in dc.get_tokens()}\n",
    "                 and len(dc.get_tokens()) < 10]\n",
    "\n",
    "samples_after_category = {}\n",
    "for ct in SAMPLE_CATEGORIES:\n",
    "    samples_after_category[ct] = defaultdict(list)\n",
    "\n",
    "for dc in disc_concepts:\n",
    "    concept_tokens = dc.get_concept_tokens()\n",
    "    all_tokens = dc.get_tokens()\n",
    "    skip_tokens = [t if t in concept_tokens  # actual token\n",
    "                   else anno.Token(None, (-1, -1), '#')  # skipped token\n",
    "                   for t in all_tokens]\n",
    "\n",
    "    cross_count_type = []\n",
    "    cross_count_example = []\n",
    "\n",
    "    concept_sequence = make_sequence(concept_tokens, CONCEPT_SEQ_COLLAPSERS)\n",
    "    concept_text = dc.get_concept()\n",
    "    samples_after_category['concept'][concept_sequence].append(dc)\n",
    "    if 'concept' in CROSS_COUNTING:\n",
    "        cross_count_type.append(concept_sequence)\n",
    "        cross_count_example.append(dc)\n",
    "\n",
    "    super_sequence = make_sequence(all_tokens, SUPER_SEQ_COLLAPSERS)\n",
    "    super_text = dc.get_covered_text()\n",
    "    samples_after_category['super'][super_sequence].append(dc)\n",
    "    if 'super' in CROSS_COUNTING:\n",
    "        cross_count_type.append(super_sequence)\n",
    "        cross_count_example.append(dc)\n",
    "\n",
    "    skip_sequence = make_sequence(skip_tokens)\n",
    "    samples_after_category['skip'][skip_sequence].append(dc)\n",
    "    if 'skip' in CROSS_COUNTING:\n",
    "        cross_count_type.append(skip_sequence)\n",
    "        cross_count_example.append(dc)\n",
    "\n",
    "    samples_after_category['cross-count'][tuple(cross_count_type)].append(\n",
    "        tuple(cross_count_example)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dict = {'type': [], 'count': [], '%': [], 'example': []}\n",
    "for type_, items in samples_after_category['super'].items():\n",
    "    data_dict['type'].append(type_)\n",
    "    data_dict['count'].append(len(items))\n",
    "    data_dict['%'].append(round(len(items) / len(disc_concepts) * 100, 2))\n",
    "    data_dict['example'].append(items[0].get_covered_text())\n",
    "    \n",
    "df = pd.DataFrame(data_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>count</th>\n",
       "      <th>%</th>\n",
       "      <th>example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[an]c[an]n</td>\n",
       "      <td>407</td>\n",
       "      <td>29.56</td>\n",
       "      <td>FY*A and FY*B alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[an]c[an][an]+n</td>\n",
       "      <td>230</td>\n",
       "      <td>16.70</td>\n",
       "      <td>Egr-1 and c-fos mRNA expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ncfan</td>\n",
       "      <td>45</td>\n",
       "      <td>3.27</td>\n",
       "      <td>TNF-alpha- or PMA-induced stimulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>([an],)+[an],?c[an]n</td>\n",
       "      <td>37</td>\n",
       "      <td>2.69</td>\n",
       "      <td>STAT6, STAT3 and STAT5 proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>nncnn</td>\n",
       "      <td>35</td>\n",
       "      <td>2.54</td>\n",
       "      <td>PTF gamma and delta subunits</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>ancnn</td>\n",
       "      <td>19</td>\n",
       "      <td>1.38</td>\n",
       "      <td>human T and B lymphocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>[an]ncn</td>\n",
       "      <td>16</td>\n",
       "      <td>1.16</td>\n",
       "      <td>DP alpha and -beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ncfann</td>\n",
       "      <td>16</td>\n",
       "      <td>1.16</td>\n",
       "      <td>N- and C- terminal SH2 domains</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ndcd</td>\n",
       "      <td>15</td>\n",
       "      <td>1.09</td>\n",
       "      <td>positions -585 and -650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>[an]+[an]ncn</td>\n",
       "      <td>15</td>\n",
       "      <td>1.09</td>\n",
       "      <td>B cell proliferation and differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>n,cnn</td>\n",
       "      <td>13</td>\n",
       "      <td>0.94</td>\n",
       "      <td>Th1, and Th2 clones</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>aacan</td>\n",
       "      <td>12</td>\n",
       "      <td>0.87</td>\n",
       "      <td>human inflammatory and immunological diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>n,cnnn</td>\n",
       "      <td>12</td>\n",
       "      <td>0.87</td>\n",
       "      <td>P2, and P4 NFAT elements</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>aancnn</td>\n",
       "      <td>12</td>\n",
       "      <td>0.87</td>\n",
       "      <td>preactivated primary NK and NK3.3 cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ncfn</td>\n",
       "      <td>11</td>\n",
       "      <td>0.80</td>\n",
       "      <td>sIgM- and sIgD-crosslinking</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ancnnn</td>\n",
       "      <td>10</td>\n",
       "      <td>0.73</td>\n",
       "      <td>human T or B cell activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>n,n,cnnn</td>\n",
       "      <td>10</td>\n",
       "      <td>0.73</td>\n",
       "      <td>P1, P2, and P4 NFAT elements</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ncfnn</td>\n",
       "      <td>8</td>\n",
       "      <td>0.58</td>\n",
       "      <td>B- and T-cell development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nncnnn</td>\n",
       "      <td>8</td>\n",
       "      <td>0.58</td>\n",
       "      <td>DR alpha and DP alpha element</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>a,can</td>\n",
       "      <td>7</td>\n",
       "      <td>0.51</td>\n",
       "      <td>infectious, or neoplastic diseases</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     type  count      %  \\\n",
       "0              [an]c[an]n    407  29.56   \n",
       "3         [an]c[an][an]+n    230  16.70   \n",
       "15                  ncfan     45   3.27   \n",
       "39   ([an],)+[an],?c[an]n     37   2.69   \n",
       "12                  nncnn     35   2.54   \n",
       "73                  ancnn     19   1.38   \n",
       "7                 [an]ncn     16   1.16   \n",
       "14                 ncfann     16   1.16   \n",
       "21                   ndcd     15   1.09   \n",
       "13           [an]+[an]ncn     15   1.09   \n",
       "45                  n,cnn     13   0.94   \n",
       "19                  aacan     12   0.87   \n",
       "2                  n,cnnn     12   0.87   \n",
       "117                aancnn     12   0.87   \n",
       "4                    ncfn     11   0.80   \n",
       "78                 ancnnn     10   0.73   \n",
       "1                n,n,cnnn     10   0.73   \n",
       "26                  ncfnn      8   0.58   \n",
       "9                  nncnnn      8   0.58   \n",
       "143                 a,can      7   0.51   \n",
       "\n",
       "                                           example  \n",
       "0                            FY*A and FY*B alleles  \n",
       "3                  Egr-1 and c-fos mRNA expression  \n",
       "15           TNF-alpha- or PMA-induced stimulation  \n",
       "39                 STAT6, STAT3 and STAT5 proteins  \n",
       "12                    PTF gamma and delta subunits  \n",
       "73                       human T and B lymphocytes  \n",
       "7                               DP alpha and -beta  \n",
       "14                  N- and C- terminal SH2 domains  \n",
       "21                         positions -585 and -650  \n",
       "13        B cell proliferation and differentiation  \n",
       "45                             Th1, and Th2 clones  \n",
       "19   human inflammatory and immunological diseases  \n",
       "2                         P2, and P4 NFAT elements  \n",
       "117        preactivated primary NK and NK3.3 cells  \n",
       "4                      sIgM- and sIgD-crosslinking  \n",
       "78                    human T or B cell activation  \n",
       "1                     P1, P2, and P4 NFAT elements  \n",
       "26                       B- and T-cell development  \n",
       "9                    DR alpha and DP alpha element  \n",
       "143             infectious, or neoplastic diseases  "
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_values('count', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B cell proliferation and differentiation\n",
      "human glial cells and lymphocytes\n",
      "PKC isoenzymes alpha and beta\n",
      "T cell activation and growth\n",
      "human GM-CSF promoter and enhancer\n",
      "T lymphocyte differentiation and activation\n",
      "T lymphocyte activation and mitogenesis\n",
      "T cell activation and differentiation\n",
      "viral gene expression and replication\n",
      "viral transcription initiation and elongation\n",
      "human chromosomes 11p15 and 11p13\n",
      "HTLV-I basal transcription and expression\n",
      "human B cell proliferation and differentiation\n",
      "IL-4 receptor subunits alpha and gamma\n",
      "T cell receptor alpha and -beta\n"
     ]
    }
   ],
   "source": [
    "for item in samples_after_category['super']['[an]+[an]ncn']:\n",
    "    print(*(t.get_covered_text() for t in item.get_tokens()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>struct</th>\n",
       "      <th>pos-seq</th>\n",
       "      <th>count</th>\n",
       "      <th>%</th>\n",
       "      <th>example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>(n n n c n)</td>\n",
       "      <td>nnncn</td>\n",
       "      <td>2</td>\n",
       "      <td>0.14</td>\n",
       "      <td>B cell proliferation and differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>(a (a n) c n)</td>\n",
       "      <td>aancn</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>human glial cells and lymphocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>(n n (n c n))</td>\n",
       "      <td>nnncn</td>\n",
       "      <td>3</td>\n",
       "      <td>0.21</td>\n",
       "      <td>PKC isoenzymes alpha and beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>(a n n c n)</td>\n",
       "      <td>anncn</td>\n",
       "      <td>2</td>\n",
       "      <td>0.14</td>\n",
       "      <td>viral gene expression and replication</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>(a n (n c n))</td>\n",
       "      <td>anncn</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>human chromosomes 11p15 and 11p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>(n ((a n) c n))</td>\n",
       "      <td>nancn</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>HTLV-I basal transcription and expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>(a n n (n c n))</td>\n",
       "      <td>annncn</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>human B cell proliferation and differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>(n n n n c n)</td>\n",
       "      <td>nnnncn</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07</td>\n",
       "      <td>T cell receptor alpha and -beta</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              struct pos-seq  count     %  \\\n",
       "26       (n n n c n)   nnncn      2  0.14   \n",
       "88     (a (a n) c n)   aancn      1  0.07   \n",
       "93     (n n (n c n))   nnncn      3  0.21   \n",
       "272      (a n n c n)   anncn      2  0.14   \n",
       "328    (a n (n c n))   anncn      1  0.07   \n",
       "409  (n ((a n) c n))   nancn      1  0.07   \n",
       "411  (a n n (n c n))  annncn      1  0.07   \n",
       "525    (n n n n c n)  nnnncn      1  0.07   \n",
       "\n",
       "                                            example  \n",
       "26         B cell proliferation and differentiation  \n",
       "88                human glial cells and lymphocytes  \n",
       "93                    PKC isoenzymes alpha and beta  \n",
       "272           viral gene expression and replication  \n",
       "328               human chromosomes 11p15 and 11p13  \n",
       "409       HTLV-I basal transcription and expression  \n",
       "411  human B cell proliferation and differentiation  \n",
       "525                 T cell receptor alpha and -beta  "
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for exp case: multi-word head: for multiple tokens after the conjunction,\n",
    "# do the tokens go with the head or the second modifier?\n",
    "df = dfs['col_full']\n",
    "subset = df[df['pos-seq'].str.contains('^[an]+[an]ncn$')]\n",
    "subset.head(25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
